keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,134 @@
+- If the fieldManager option is set for the request (for example through kubectl), it will get used as manager
+- If the newly introduced `fieldManager` field is set in the received request, it will get used as manager
+- If both of the  two options are set and do not match, the request will get rejected.
+
+An example of setting the field would be:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: example
+  fieldManager: jenkins
+data:
+  k: v
+```
+
+### User Stories [optional]
+
+#### Git Managed Configuration
+
+Alice has a repo containing her current set of Kubernetes manifests as yaml in git.
+The repo gets managed using Gitflow.
+Master branch is automatically applied to the production Kubernetes cluster by a push triggered Jenkins job.
+Staging branch is automatically applied to the staging Kubernetes cluster by a push triggered Jenkins job.
+All members of her team file pull requests against the repo to bring changes into staging and production.
+
+Aside from that, Alice and her team got services that directly manage their objects, like controllers and webhooks,
+as well as services that updated labels for some objects.
+
+They want to make use of the new Server Side Apply, so concurrent changes to objects from different actors are possible without unwanted overlap/conflicts.
+
+To do so, assuming the changes in this KEP, Alice would add the newly introduced field to all manifests in their config management repo. The field would be set to e.g. `configManagement` (or their repo name).
+Additionally, she would update the manifests sent by different actors (like controllers) to contain the field with a value like `controllerName`. Note that this might already be the case for some through the defaulting to the user-agent inside the apply and update pathway.
+
+If now the team updates their repo with a change to their manifests, Jenkins again applies them as usual. On the apiserver side though, the fieldManager set in the applied manifest gets used and all fields set now get owned by the `configManagement` fieldManager.
+
+Again as usual other actors step in to change the objects. They as well use their fieldManager value either set in the manifest file they apply, or based on their user-agent/configuration.
+One of those actors is setting annotations for every service with external information.
+As the actor now only updates annotations not already owned by the `configManagement` fieldManager, applying them will cause the actor to own them.
+
+All actors now can coexist and manage their set of fields without interfering with each other.
+
+### Risks and Mitigations
+
+One risk might be, that introducing this change would mean there is a field that does not get persisted to storage.
+As a result, setting it results in different behavior on the apiserver, but does not reflect when reding the object back from the apiserver.
+
+This can seem unintuitive and we need to make sure documentation is right on this.
+
+## Design Details
```

Comment:
It seems like there's two ways to make this work.

1. kube-apiserver reads the field, stores that as the manager, then clears the field before proceeding.
2. kube-apiserver always clears the field and ignores it (or even rejects requests setting it). kubectl reads the field and puts it into the query parameter.

Either way it should always be an error to set this field on a POST, PUT, or non-apply PATCH request.

Option 2 is slightly more work for implementers. I think this document should say which option is chosen. I tentatively vote for option 1.

Reply:
Current implementation is Option 1

Reply:
> Either way it should always be an error to set this field on a POST, PUT, or non-apply PATCH request.

Why? POST and PUT take ownership of fields as well?

Reply:
I guess as long as the field is not actually stored (proving that if the client set it, they must have done so deliberately), then it would be OK to honor it in all requests?